core/src/main/java/lucee/runtime/sql/QueryPartitions.java

```diff
@@ -0,0 +1,311 @@
+			// Populate additional columns needed for operations but are not found in the select
+			// list above
+			for (Collection.Key col: additionalColumns) {
+				if (source.containsKey(col)) {
+					targetPartition.setAt(col, targetPartition.getRecordcount(), source.getAt(col, row, null), true);
+				}
+			}
+		}
+	}
+
+	/**
+	 * Generate a unique string that represents the column data being grouped on
+	 * 
+	 * @param pc PageContext
+	 * @param source QueryImpl to get data from. Note, operations have not yet been processed
+	 * @param row Row to get data from
+	 * @param finalizedColumnVals If we're adding finalized data, just copy it across. Easy. This
+	 *            applies when distincting a result set after it's already been processed
+	 * @return unique string
+	 * @throws PageException
+	 */
+	public String buildPartitionKey(PageContext pc, Query source, int row, boolean finalizedColumnVals) throws PageException {
+		String partitionKey = "";
+		for (int cell = 0; cell < groupbys.length; cell++) {
+			String value;
+			// This is when reading columns out of a previous union query that doesn't have any
+			// expressions in it, just literal values It's important that we are just getting this
+			// value by index since the group by expressions may be a reference to the select
+			// expressions from another query object
+			if (finalizedColumnVals) {
+				value = Caster.toString(source.getAt(source.getColumnNames()[cell], row));
+			}
+
+			else {
+				value = Caster.toString(qoQ.getValue(pc, sql, source, row, null, groupbys[cell]));
+			}
+			// Internally Java uses a StringBuilder for this concatenation
+			partitionKey += createUniqueValue(value, groupbys[cell].toString(false));
+		}
+		return partitionKey;
+	}
+
+	/**
+	 * Helper function to turn column data into string
+	 * 
+	 * @param value
+	 * @param col
+	 * @return
+	 * @throws PageException
+	 */
+	private String createUniqueValue(String value, String col) throws PageException {
+
+		// There doesn't seem to be a key length on a HashMap, but it seems like a good
+		// idea to hash long values. Not hashing everything, because that is slower.
+		if (value.length() > 255) {
+			try {
+				return MD5.getDigestAsString(value);
+			}
+			catch (IOException e) {
+				throw new DatabaseException("Unable to hash query value for column [" + col + "] for partitioning.", e.getMessage(), null, null);
+			}
+		}
+		else {
+			// Inject some characters to prevent accidental overlap of data been nearby columns
+			return "______________" + value;
+		}
+	}
+
+	/**
+	 * Get number of partitions
+	 * 
+	 * @return
+	 */
+	public int getPartitionCount() {
+		return partitions.size();
+	}
+
+	/**
+	 * Get partition Map
+	 * 
+	 * @return
+	 */
+	public HashMap<String, QueryImpl> getPartitions() {
+		return partitions;
+	}
+
+	/**
+	 * Get array of grouped QueryImpl object
+	 * 
+	 * @return
+	 */
+	public QueryImpl[] getPartitionArray() {
+		return (QueryImpl[]) partitions.values().toArray();
```

Comment:
i see in a lot of places "QueryImpl" is used, even "Query" is sufficient, whenever possible go down the the interface that is sufficient. by using "Query" you make the code more flexible.

Reply:
@michaeloffner Yes, I am aware that you always use the interface.  however, the issue I was running into is that not all the methods declared in `QueryImpl` are also declared in the `Query` interface.  `cutRowsTo()` is an example as well as the new overloaded `setAt()` method I added.  I understand changing the interface in a minor version is not an option so I did not touch the interface.  

Since you brought it up, I have changed all the types over to use the `Query` interface and I am casting to `QueryImpl` in the places it is required.  Ex:
```java
((QueryImpl) target).cutRowsTo(maxrows);
```
Is that an ok workaround?